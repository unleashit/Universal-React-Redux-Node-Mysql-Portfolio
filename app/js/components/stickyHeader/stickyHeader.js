import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Navigation from '../navigation/navigation';
import OpenChat from '../ReactHelpDesk/client/openChat';
import LiveChatStatus from '../ReactHelpDesk/client/liveChatStatus';
import * as globalActions from '../../actions/global';
import throttle from 'lodash/throttle';

export class StickyHeader extends React.Component {
    constructor(props) {
        super(props);
        this.boundHandleScroll = throttle(
            this.handleScroll.bind(this, props.dispatch),
            150
        );
    }

    componentDidMount() {
        window.addEventListener('scroll', this.boundHandleScroll);
    }

    componentWillUnmount() {
        window.removeEventListener('scroll', this.boundHandleScroll);
        document.documentElement.className = '';
        const { dispatch, global } = this.props;
        if (global.headerState) dispatch(globalActions.setHeader(false));
    }

    handleScroll() {
        const { dispatch } = this.props;
        const scrl = window.pageYOffset;

        // handle sticky header
        if (scrl >= 250 && this.props.global.headerState === false) {
            document.documentElement.className = 'sticky-menu-open';
            dispatch(globalActions.setHeader(true));
        } else if (scrl < 250 && this.props.global.headerState === true) {
            document.documentElement.className = '';
            dispatch(globalActions.setHeader(false));
        }
    }

    render() {
        const {
            visible,
            global: { headerState },
        } = this.props;

        const classes = `sticky-header hidden-xs-down${
            visible || headerState ? ' on' : ' off'
        }`;

        return (
            <div className={classes}>
                <div className="container-fluid">
                    <Navigation ulClass="sticky-nav" home={true} logo="1" />
                    <div
                        className="hamburger-alt hidden-md-up pull-right"
                        onClick={this.props.openBurger}
                    >
                        <i className="fa fa-bars" /> &nbsp;MENU
                    </div>
                    <div className="contact-info pull-right hidden-xs-down">
                        <div>
                            <span className="chat-status">
                                <OpenChat>
                                    <LiveChatStatus
                                        remoteId={this.props.remoteId}
                                    />
                                </OpenChat>
                            </span>
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}

StickyHeader.propTypes = {
    visible: PropTypes.bool,
    openBurger: PropTypes.func.isRequired,
    remoteId: PropTypes.string.isRequired,
};

/* istanbul ignore next */
function mapStateToProps(state) {
    return {
        global: state.global,
    };
}
/* istanbul ignore next */
function mapDispatchToProps(dispatch) {
    return {
        dispatch: dispatch,
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(StickyHeader);
