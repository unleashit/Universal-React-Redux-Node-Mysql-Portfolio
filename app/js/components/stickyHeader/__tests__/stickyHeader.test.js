import { StickyHeader } from '../stickyHeader';
import { Footer } from '../../footer/footer';
import * as utils from '../../../libs/utils';

describe('<StickyHeader />', () => {
    let wrapper;
    let props;

    beforeEach(() => {
        props = {
            visible: false,
            openBurger: jest.fn(),
            remoteId: 'admin',
            dispatch: jest.fn(),
            global: {
                headerState: false
            }
        };
        wrapper = shallow(<StickyHeader {...props} />);
    });

    test('renders without crashing', () => {
        expect(wrapper.find('.sticky-header')).toHaveLength(1);
    });

    test('snapshot matches', () => {
        expect(wrapper).toMatchSnapshot();
    });

    describe('lifecycle methods', () => {
        test('componentDidMount()', async () => {
            props.dispatch.mockReset();
            const mockAddEvent = jest.spyOn(window, 'addEventListener');
            await wrapper.instance().componentDidMount();

            expect(mockAddEvent).toHaveBeenCalledWith(
                'scroll',
                expect.any(Function)
            );
        });

        test('componentWillUnmount()', () => {
            props.global.headerState = true;
            wrapper = shallow(<StickyHeader {...props} />);

            const mockRemoveEvent = jest.spyOn(window, 'removeEventListener');
            props.dispatch.mockReset();
            wrapper.instance().componentWillUnmount();

            expect(mockRemoveEvent).toHaveBeenCalledWith(
                'scroll',
                expect.any(Function)
            );
            expect(document.documentElement.className).toEqual('');
            expect(props.dispatch).toHaveBeenCalledTimes(1);
        });
    });

    test('displays `on` or `off` class depending on `visible` prop ', () => {
        expect(wrapper.find('.sticky-header.off')).toHaveLength(1);
        expect(wrapper.find('.sticky-header.on')).toHaveLength(0);
        wrapper.setProps({ visible: true });
        expect(wrapper.find('.sticky-header.off')).toHaveLength(0);
        expect(wrapper.find('.sticky-header.on')).toHaveLength(1);
    });

    describe('handleScroll()', () => {
        test('sets sticky header state to `off` when page is not scrolled', () => {
            props.global.headerState = true;
            wrapper = shallow(<StickyHeader {...props} />);
            props.dispatch.mockReset();

            window.pageYOffset = 0;
            wrapper.instance().handleScroll();

            expect(props.dispatch).toHaveBeenCalledTimes(1);
            expect(document.documentElement.className).toEqual('');
        });

        test('sets sticky header state to `on` when page is scrolled', () => {
            props.dispatch.mockReset();
            window.pageYOffset = 251;
            wrapper.instance().handleScroll();

            expect(props.dispatch).toHaveBeenCalledTimes(1);
            expect(document.documentElement.className).toEqual(
                'sticky-menu-open'
            );
        });
    });

    // test('analytics hooks are called', () => {
    //     jest.spyOn(StickyHeader.prototype, 'analytics');
    //     const mockReactGA = jest
    //         .spyOn(ReactGA, 'event')
    //         .mockImplementation(() => {});
    //
    //     wrapper = shallow(<StickyHeader {...props} />);
    //
    //     const chatStatus = wrapper.find('.chat-status');
    //     chatStatus.simulate('click');
    //     expect(wrapper.instance().analytics).toHaveBeenCalledWith(
    //         'chat in header'
    //     );
    //
    //     expect(mockReactGA).toHaveBeenCalledTimes(1);
    // });
});
