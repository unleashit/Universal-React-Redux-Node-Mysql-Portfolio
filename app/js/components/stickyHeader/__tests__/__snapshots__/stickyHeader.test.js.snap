// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StickyHeader /> snapshot matches 1`] = `
<div
  className="sticky-header hidden-xs-down off"
>
  <div
    className="container-fluid"
  >
    <Navigation
      home={true}
      logo="1"
      ulClass="sticky-nav"
    />
    <div
      className="hamburger-alt hidden-md-up pull-right"
      onClick={[MockFunction]}
    >
      <i
        className="fa fa-bars"
      />
       Â MENU
    </div>
    <div
      className="contact-info pull-right hidden-xs-down"
    >
      <div>
        <span
          className="chat-status"
        >
          <Connect(OpenChat)>
            <liveChatStatus
              remoteId="admin"
            />
          </Connect(OpenChat)>
        </span>
      </div>
    </div>
  </div>
</div>
`;
